kliff/uq/mcmc.py

```diff
@@ -135,237 +140,218 @@ def __new__(
+            sampling. It defaults to the natural temperature :math:`T_0` [Frederiksen2004]_.
+        logprior_fn : Optional[Callable]
+            A function that evaluate logarithm of the prior
+            distribution. The prior doesn't need to be normalized. It
+            defaults to a uniform prior over a finite range.
+        logprior_args : Optional[tuple]
+            Additional positional arguments of the ``logprior_fn``. If the
+            default ``logprior_fn`` is used, then the boundaries of the
+            uniform prior can be specified here.
+        **kwargs : Optional[dict]
+            Additional keyword arguments for ``emcee.EnsembleSampler``.
+
+        Attributes
+        ----------
+        loss: Loss
+            Loss function instance from :class:`~kliff.loss.Loss`
+        T: float
+            Values of the sampling temperature.
+        sampler: emcee.EnsembleSampler
+            Sampler instance from ``emcee.EnsembleSampler``
+
+        Notes
+        -----
+        As a convention, KLIFF inflates the likelihood by some sampling temperature, i.e.,
+        :math:`L(\\theta) \propto \exp(-C(\\theta) / T)`. As a default, the sampling
+        temperature is set to the natural temperature. To use the untempered likelihood
+        (:math:`T=1`), user should specify the argument ``T=1``.
+
+
+        References
+        ----------
+        .. [Frederiksen2004] S. L. Frederiksen, K. W. Jacobsen, K. S. Brown, and J. P.
+           Sethna, “Bayesian Ensemble Approach to Error Estimation of Interatomic
+           Potentials,” Phys. Rev. Lett., vol. 93, no. 16, p. 165501, Oct. 2004,
+           doi: 10.1103/PhysRevLett.93.165501.
+        """
 
-        # Probability
-        if T is None:
-            self.T = get_T0(self.loss)
-        else:
-            self.T = T
-        logl_fn = self._loglikelihood_wrapper
-        logp_fn = self._logprior_wrapper(logprior_fn, *logprior_args)
+        def __init__(
+            self,
+            loss: Loss,
+            nwalkers: Optional[int] = None,
+            T: Optional[float] = None,
+            logprior_fn: Optional[Callable] = None,
+            logprior_args: Optional[tuple] = None,
+            **kwargs,
+        ):
+            self.loss = loss
 
-        global logprobability_fn
+            # Dimensionality
+            ndim = loss.calculator.get_num_opt_params()
+            nwalkers = 2 * ndim if nwalkers is None else nwalkers
 
-        def logprobability_fn(x):
-            return logl_fn(x) + logp_fn(x)
+            # Probability
+            if T is None:
+                self.T = get_T0(self.loss)
+            else:
+                self.T = T
+            logl_fn = self._loglikelihood_wrapper
+            logp_fn = self._logprior_wrapper(logprior_fn, *logprior_args)
 
-        self.sampler = emcee.EnsembleSampler(
-            nwalkers, ndim, logprobability_fn, **kwargs
-        )
+            global logprobability_fn
```

Comment:
Do we have to use global function? If possible, can this function be defined outside the class and used here? 

Also, it worries me that we have two global functions with the same name `logprobability_fn` here and above. Will they clash with each other?

Reply:
I use global function because otherwise I always got the following error:
```
TypeError: cannot pickle 'kimpy.model.PyModel' object
```
Do you have any suggestion to fix this?

Reply:
@mjwen The `PtemceeSampler` class uses `loglikelihood_fn` and the `EmceeSampler` class uses `logprobability_fn`. So they won't clash with each other.

And from what I tried, we can avoid declaring a global variable if we define `logprobability_fn` before instantiating `kliff.uq.MCMC`, and pass in that function as an argument of this class. However, it seems to defeat the purpose of this package if we require user to define this function manually. Instead, what if we add an optional argument to `kliff.uq.MCMC` to allow user to pass in their own `logprobability_fn` and have a default value with an if-else statement like the following:
```
if logprobability_fn is None:
    global logprob_fn
    def logprob_fn:
        ...
else:
    logprob_fn = logprobability_fn
```
Unfortunately, I still don't know how we can get rid of the global variable completely.